### DISCUSSION TITLE
Jobrunr and Spring Boot integration test caching

### DISCUSSION THREAD

**Discussion by wimdeblauwe** [jobrunr/jobrunr#908] [General]
We had an issue with a random failing integration test in our application that uses Spring Boot and Jobrunr. I managed to reproduce the issue in a small sample application and found out that the combination of Spring Boot, Testcontainers and Jobrunr leads to this issue.

I don't think there is really a bug here, it is just something people might run into as well, so I wanted to document it here. Feel free to add something to the FAQ on the Jobrunr website about it if you want.

The issue is that a job may be scheduled within a certain Spring application context that has a certain set of beans, but that the execution of the job happens within another Spring application context that has a different set of beans.

## Background info

To understand why this happens, you need some background info:

When you use Spring Boot and Testcontainers, you have 2 main ways to start the docker containers:

1) Use `@TestContainers` and `@Container`. Together with `@ServiceConnection`, it makes it real easy to set everything up:

```
@SpringBootTest
@Testcontainers
class JobrunrIntegrationTestApplicationTests {

  @Container
  @ServiceConnection
  static PostgreSQLContainer<?> postgresContainer = new PostgreSQLContainer<>(DockerImageName.parse("postgres:latest"));

  @Autowired
  private ExampleService exampleService;

  @Test
  void contextLoads() {
    await()
        .untilAsserted(() -> {
          assertThat(exampleService.wasServiceMethodCalled()).isTrue();
        });
  }
}
```

2) Use an `ApplicationContextInitializer` where start the Docker image via a static block:

```
public class MyContextInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

  static PostgreSQLContainer<?> postgresContainer = new PostgreSQLContainer<>(DockerImageName.parse("postgres:latest"));

  static {
    postgresContainer.start();
  }

  @Override
  public void initialize(ConfigurableApplicationContext applicationContext) {
    TestPropertyValues.of(
        "spring.datasource.url=" + postgresContainer.getJdbcUrl(),
        "spring.datasource.username=" + postgresContainer.getUsername(),
        "spring.datasource.password=" + postgresContainer.getPassword()
    ).applyTo(applicationContext.getEnvironment());
  }
}
```

The advantage of the second method is that there is only 1 instance of the Docker container that is re-used for all tests. For a big test suite, this will significantly increase the execution speed of your tests.

Another piece of the puzzle is that Spring test framework uses context caching to speed up test execution. See https://rieckpil.de/improve-build-times-with-context-caching-from-spring-test/ for more background information on that mechanism. One of the ways that Spring decides that a new test context should be created is if you have different profiles active on different tests.

## Reproducer

To reproduce the issue, I created 2 services: `ExampleService` and `ExampleService2`. The latter is only enabled when proflie `service2` is active (Via `@Profile("service2")`).

Then 2 command line runners will each schedule a job, using either service:

```
@Component
public class InitJobCommandLineRunner implements CommandLineRunner {

  private final JobScheduler scheduler;
  private final ExampleService exampleService;

  public InitJobCommandLineRunner(JobScheduler scheduler, ExampleService exampleService) {
    this.scheduler = scheduler;
    this.exampleService = exampleService;
  }

  @Override
  public void run(String... args) {
    scheduler.create(JobBuilder.aJob()
        .withName("my-job")
        .scheduleIn(Duration.ofSeconds(30))
        .withDetails(exampleService::doIt));
  }
}
```

and 

```
@Component
@Profile("service2")
public class InitJobCommandLineRunner2 implements CommandLineRunner {

  private final JobScheduler scheduler;
  private final ExampleService2 exampleService;

  public InitJobCommandLineRunner2(JobScheduler scheduler, ExampleService2 exampleService) {
    this.scheduler = scheduler;
    this.exampleService = exampleService;
  }

  @Override
  public void run(String... args) {
    scheduler.create(JobBuilder.aJob()
        .withName("my-job-2")
        .scheduleIn(Duration.ofSeconds(30))
        .withDetails(exampleService::doIt));
  }
}
```

I then have 4 almost identical tests. Two don't activate profile, and two others activate the `service2` profile. When the tests are written using the `@Container` annotation, there is no problem. However, when the tests are written using the static initializer block, then random failures appear.

This is what happens if you run the tests:

1. Test1 runs and starts the docker container. Spring Boot starts and everything that is in the context is cached. Jobrunr starts and schedules the job. The background service picks up the job and executes it. The test is ok.
2. Test2 (which has the `service2` profile) runs. Spring Boot test framework notices that there should be a new context started since this test has an extra profile active. Note that all the Spring beans are started a second time, so also all the Jobrunr related beans. The jobs are scheduled, but there are now 2 instances of the background service both reading from the same database. It can now happen that the context that was created for Test1 picks up the job and tries to execute it.  However, that context does not have ExampleService2 in its context. So there is an error `No qualifying bean of type available` and the test fails:

```
org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.example.jobrunrintegrationtest.ExampleService2' available
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:343) ~[spring-beans-6.1.2.jar:6.1.2]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:334) ~[spring-beans-6.1.2.jar:6.1.2]
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1250) ~[spring-context-6.1.2.jar:6.1.2]
	at org.jobrunr.server.runner.BackgroundJobWithIocRunner.supports(BackgroundJobWithIocRunner.java:21) ~[jobrunr-6.3.3.jar:6.3.3]
	at org.jobrunr.server.BackgroundJobServer.lambda$getBackgroundJobRunner$0(BackgroundJobServer.java:218) ~[jobrunr-6.3.3.jar:6.3.3]
```

## Conclusion

To avoid the problem either:

* Use `@Container` instead of the static initializer trick (But I don't really like the increased time this adds to your test suite)
* Ensure that all integration tests that need Jobrunr are using the same Spring test context.

**Comment by rdehuyss** [comment]
I wonder if there is no shutdown hook we can use (to gracefully shutdown the beans from the first context) when the second context starts?

**Reply by wimdeblauwe** [reply]
That would not help as there might be other tests that will use that first context. Unless, there would also be some kind of hook to when a cached context is taken from the cache to activate the beans again?
